ihmc-robotics-toolkit/src/main/java/us/ihmc/robotics/math/filters/TransferFunctionDiscretizer.java

```diff
@@ -0,0 +1,204 @@
+package us.ihmc.robotics.math.filters;
+
+import org.ejml.data.DMatrixRMaj;
+import org.ejml.dense.row.CommonOps_DDRM;
+
+/**
+ * This class is intended to be used with the YoFilteredDouble object to filter a signal with a desired causal transfer function.
+ * The TransferFunctionDiscretizer creates a discrete filter from a continuous transfer function representation:
+ * <pre>
+ * G(s) = k * (b_0*s^m + b_1*s^{m-1} + ... + b_{m-1}*s + b_m)/(a_0*s^n + a_1*s^{n-1} + ... + a_{n-1}*s + a_n)
+ * </pre>
+ * Where n >= m for the transfer function to be causal. This class does not check whether your desired transfer function is stable.
+ * An example simulation of this class can be found in the YoFilteredDoubleSimulation.java class. More details can be found in the
+ * associated paper: "Software Implementation of Digital Filtering via Tustin's Bilinear Transform" by Connor Herron.
+ *
+ * @author Connor Herron
+ */
+public class TransferFunctionDiscretizer
+{
+
+   // Used for temporary private calculations.
+   private DMatrixRMaj outputVec;
+   private DMatrixRMaj tempVec;
+   private int numCols;
+
+   // Desired Transfer function numerator and denominator.
+   private DMatrixRMaj numerator;
+   private DMatrixRMaj denominator;
+
+   // Input/Output Coefficients.
+   private DMatrixRMaj inputCoefficients;
+   private DMatrixRMaj outputCoefficients;
+
+   private double k;
+   private double p;
+   private double fs;
+   private double output;
+   private int n;
+   private int m;
+   private String name;
+
+   public TransferFunctionDiscretizer(ContinuousTransferFunction tf, double sampleFrequency)
+   {
+      this(tf.getName(), tf, sampleFrequency);
+   }
+
+   /*
+    * @param name Name of the filter.
+    * @param k Transfer Function Gain
+    * @param numerator Order of continuous polynomial coefficients from largest to smallest order, i.e. [b_0, b_1, ... , b_{m-1}, b_{m}].
+    * @param demoninator Order of continuous polynomial coefficients from largest to smallest order, i.e. [a_0, a_1, ..., a_{n-1}, a_n].
+    * @param sampleFrequency Frequency [Hz] which this variable will be updated. Filter performance will be worse if this is not accurate.
+    */
+   public TransferFunctionDiscretizer(String name, ContinuousTransferFunction tf, double sampleFrequency)
+   {
+
+      this.name = name;
+
+      m = tf.getNumerator().length;
+      n = tf.getDenominator().length;

```

Comment:
these may need to change, or be direct methods (`getNumberOfPoles()`, `getNumberOfZeroes()`)